File: vuln-sandbox/ruby/cwe-22/path-traversal.rb
# CWE-22: Path Traversal via File.read in Ruby
# Rule: ruby-path-traversal-file-read
# Expected: TruePositive

class DocumentsController < ApplicationController
  def show
    # VULNERABLE: User-controlled filename used in File.read
    filename = params[:filename]
    content = File.read("/uploads/#{filename}")
    render plain: content
  end

  def download
    # VULNERABLE: User input in File.open
    path = params[:path]
    File.open("/data/#{path}") do |f|
      send_data f.read
    end
  end
end

File: vuln-sandbox/php/cwe-22/path-traversal.php
<?php
// CWE-22: Path Traversal via include/file_get_contents in PHP
// Rule: php-path-traversal-file-get-contents
// Expected: TruePositive

// VULNERABLE: User-controlled filename in file_get_contents
$filename = $_GET['file'];
$content = file_get_contents("/uploads/$filename");
echo $content;

// VULNERABLE: User input in fopen
$path = $_REQUEST['path'];
$handle = fopen("/data/$path", 'r');
if ($handle) {
    echo fread($handle, filesize("/data/$path"));
    fclose($handle);
}
?>
